### ISSUE TITLE
numba.cuda.detect() feature request

### ISSUE THREAD

**Issue by cristipurdel** [numba/numba#3203]
- [ ] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [ ] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

I would prefer to have the following info, maybe also in the cudatoolkit, so that users with low end cards can see why 'cuda' is not so fast as expected (single vs double speed, device - host speed) or why the kernel is 'crashing' (watchdog/TDR "windows feature")

From CUDA-Z utility:
Watchdog Enabled : Yes / No
Compute Mode : Default or TCC
Performance: Host to Device; Device to Host; Host to Host; Single-precision float; Double-precision float;

Fron Nsight Monitor: WDDM TDR Delay; WDDM TDR Enabled

**Comment by seibert** [comment]
To add this to Numba, we can get some of these attributes from `cuDeviceGetAttribute`:
- `CU_DEVICE_ATTRIBUTE_TCC_DRIVER`: `1` if TCC driver is in use.  If `0` on Windows, then WDDM is in use and performance will be lower.
- `CU_DEVICE_ATTRIBUTE_SINGLE_TO_DOUBLE_PRECISION_PERF_RATIO`: Ratio of single to double precision throughput.  A large number (>3) indicates a card with slow double precision.
- `CU_DEVICE_ATTRIBUTE_KERNEL_EXEC_TIMEOUT`: `1` if there is a time limit for running kernels (i.e., the watchdog is present)

We should output these values in the `numba -s` output for diagnostics.

**Comment by rawwar** [comment]
can this issue be labelled as a "good first issue"? If there are any tricky things here, please let me know. I am working on this

**Comment by gmarkall** [comment]
@rawwar Sure - go for it!

**Comment by rawwar** [comment]
My current update outputs the following 
```
In [5]: from numba import cuda

In [6]: cuda.detect()
Found 1 CUDA devices
id 0    b'GeForce GTX 1660 Ti'                              [SUPPORTED]
                      compute capability: 7.5
                           pci device id: 0
                              pci bus id: 1
                        Watchdog Enabled: 1
                            Compute Mode: 0
Single To Double Precision Performance ratio: 32
Summary:
        1/1 devices are supported
```

I understand that the original request is to print Host to Device, Device to Host, Host to Host, FP32 and FP64 performance. 
(Please correct me if I am wrong) But, These values are dynamic(based on how busy is the GPU) and we can try to get an average value by copying data multiple times and timing it.  

Also, I am currently printing `Single To Double Precision Performance ratio` which might be too long. Please suggest a shorter one. I am thinking of "FP32 : FP64 performance". But, even I don't like it.

Regarding `WDDM TDR Delay` and `WDDM TDR Enabled`, I am not sure how to get them. Based on what i read so far, this is specific to windows and it's in registry - [Link](https://docs.microsoft.com/en-us/windows-hardware/drivers/display/tdr-registry-keys)

**Comment by gmarkall** [comment]
> My current update outputs the following
> 
> ```
> In [5]: from numba import cuda
> 
> In [6]: cuda.detect()
> Found 1 CUDA devices
> id 0    b'GeForce GTX 1660 Ti'                              [SUPPORTED]
>                       compute capability: 7.5
>                            pci device id: 0
>                               pci bus id: 1
>                         Watchdog Enabled: 1
>                             Compute Mode: 0
> Single To Double Precision Performance ratio: 32
> Summary:
>         1/1 devices are supported
> ```

This is looking nice! For "Compute Mode", could you translate 0 / 1 to the appropriate string of either "WDDM" or "TCC"?

> I understand that the original request is to print Host to Device, Device to Host, Host to Host, FP32 and FP64 performance.
> (Please correct me if I am wrong) But, These values are dynamic(based on how busy is the GPU) and we can try to get an average value by copying data multiple times and timing it.

I think for now we should just leave out the performance benchmarks, and I would suggest making a PR with what you have already.

> Also, I am currently printing `Single To Double Precision Performance ratio` which might be too long. Please suggest a shorter one. I am thinking of "FP32 : FP64 performance". But, even I don't like it.

Your suggestion of "FP32 : FP64 performance" looks good to me.

> Regarding `WDDM TDR Delay` and `WDDM TDR Enabled`, I am not sure how to get them. Based on what i read so far, this is specific to windows and it's in registry - [Link](https://docs.microsoft.com/en-us/windows-hardware/drivers/display/tdr-registry-keys)

I would suggest that knowing there is a watchdog enabled and being able to see whether the mode is WDDM or TCC is a good start - if the TDR delay is important, it could be added in a subsequent PR.

**Comment by stuartarchibald** [comment]
Implemented in #6809, closing.
